src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java

```diff
@@ -0,0 +1,185 @@
+ */
+public final class KustoClientInstance {
+
+    private static final Logger LOGGER = StatusLogger.getLogger();
+    private static final AtomicReference<KustoClientInstance> ATOMIC_INSTANCE = new AtomicReference<>();
+    private static final RetryRegistry RETRY_REGISTRY = RetryRegistry.ofDefaults();
+    private final IngestClient ingestClient;
+    private final IngestionProperties ingestionProperties;
+    Retry ingestionRetry;
+
+    private KustoClientInstance(KustoLog4jConfig kustoLog4jConfig) throws URISyntaxException {
+        // default max attempts is 3 !
+        RetryConfig retryConfig = RetryConfig.custom()
+                .intervalFunction(IntervalFunction.ofExponentialBackoff(kustoLog4jConfig.backOffMinSeconds,
+                        IntervalFunction.DEFAULT_MULTIPLIER, kustoLog4jConfig.backOffMaxSeconds))
+                .retryOnException(this::isTransientException)
+                .failAfterMaxAttempts(false)
+                .build();
+        ingestionRetry = RETRY_REGISTRY.retry(INGESTION_RETRIES, retryConfig);
+        ConnectionStringBuilder csb = "".equals(kustoLog4jConfig.appKey) || "".equals(kustoLog4jConfig.appTenant)
+                ? ConnectionStringBuilder.createWithAadManagedIdentity(kustoLog4jConfig.clusterPath,
+                        kustoLog4jConfig.appId)
+                : ConnectionStringBuilder.createWithAadApplicationCredentials(kustoLog4jConfig.clusterPath,
+                        kustoLog4jConfig.appId,
+                        kustoLog4jConfig.appKey, kustoLog4jConfig.appTenant);
+        csb.setClientVersionForTracing(String.format("Kusto.Log4j.Connector:%s", getPackageVersion()));
+        if (kustoLog4jConfig.proxyUrl != null && !kustoLog4jConfig.proxyUrl.trim().equals("")) {
+            HttpClientProperties proxy = HttpClientProperties.builder().proxy(HttpHost.create(kustoLog4jConfig.proxyUrl)).build();
+            LOGGER.warn("Using proxy : {} ", kustoLog4jConfig.proxyUrl);
+            ingestClient = IngestClientFactory.createClient(csb, proxy);
+        } else {
+            ingestClient = IngestClientFactory.createClient(csb);
+        }
+        LOGGER.info("Using database : {}  & table {} to ingest logs", kustoLog4jConfig.dbName,
+                kustoLog4jConfig.tableName);
+        ingestionProperties = new IngestionProperties(kustoLog4jConfig.dbName, kustoLog4jConfig.tableName);
+        ingestionProperties.setFlushImmediately(kustoLog4jConfig.flushImmediately);
+        if (kustoLog4jConfig.logTableMapping != null && kustoLog4jConfig.mappingType != null
+                && !"".equals(kustoLog4jConfig.logTableMapping.trim()) &&
+                !"".equals(kustoLog4jConfig.mappingType.trim())) {
+            LOGGER.error("Using mapping {}  of type  {} ", kustoLog4jConfig.logTableMapping,
+                    kustoLog4jConfig.mappingType);
+            IngestionMapping.IngestionMappingKind mappingType = "json".equalsIgnoreCase(kustoLog4jConfig.mappingType) ? JSON : CSV;

```

Comment:
Better to enrich the `IngestionMappingKind(String kustoValue)` constructor to accept can insensitive input, and use that.